cpp/src/strings/json/json_to_array.cu

```diff
@@ -0,0 +1,598 @@
+      pos++;
+    }
+    else if ((c >= '0' && c <= '9') || c == '+' || c == '-') {
+      char const* comma_pos = find_char_pos(',');
+      char const* last_value_pos = find_last_validity_pos();
+      value = string_view(pos, last_value_pos - pos);
+      pos = last_value_pos;
+      parse_whitespace();
+      if (*pos == ',' && comma_pos == pos) { //obj inner
+        finish_obj = false;
+        finish_arr = false;
+        pos++;
+      }
+      else if (*pos == '}' && eof(comma_pos)) { //eof end
+        finish_obj = true;
+        finish_arr = true;
+      }
+      else if (*pos == '}'){ // obj end
+        pos = comma_pos;
+        pos++;
+        finish_obj = true;
+        finish_arr = false;
+      }
+    }
+    else {
+      return parse_result::ERROR;
+    }
+    output.add_output({value.data(), static_cast<size_type>(value.size_bytes())});
+    output.offset[num_element++] = value.size_bytes();
+    return parse_result::SUCCESS;
+  }
+
+  first_operator_type element;
+
+  private:
+
+  const char* cur_el_start;
+  size_t num_element;
+};
+
+#define PARSE_TRY(_x)                                                       \
+  do {                                                                      \
+    last_result = _x;                                                       \
+    if (last_result == parse_result::ERROR) { return parse_result::ERROR; } \
+  } while (0)
+
+__device__ parse_result parse_json_array(json_state& j_state,
+                                                json_array_output& key_output,
+                                                json_array_output& value_output,
+                                                size_t& num_lists
+                                                )
+{
+  json_first_operator op = j_state.get_first_character();
+  switch(op.type) {
+    case OBJECT: {
+      for (size_t i = 0 ; ; i++)
+      {
+        bool finish_obj = false;
+        bool finish_arr = false; // not use in this case
+        parse_result last_result = parse_result::SUCCESS;
+        PARSE_TRY(j_state.extract_key(key_output));
+        PARSE_TRY(j_state.extract_value(value_output, finish_obj, finish_arr));
+        num_lists = 1;
+        if (finish_obj) {
+          key_output.list_offsets[0] = key_output.num_rows;
+          break;
+        }
+      } 
+    } break;
+    case ARRAY: {
+      for (size_t i = 0 ;  ; i++)
+      {
+        bool finish_arr = false;
+        size_t j = 0;
+        for ( ;  ; j++)
+        {
+          bool finish_obj = false;
+          parse_result last_result = parse_result::SUCCESS;
+          PARSE_TRY(j_state.extract_key(key_output));
+          PARSE_TRY(j_state.extract_value(value_output, finish_obj, finish_arr));
+          if (finish_obj) break;
+        }
+        num_lists++;
+        key_output.list_offsets[i] = j + 1;
+        if (finish_arr) break;
+      }
+    } break;
+    case NONE: {
+      return parse_result::ERROR;
+    } break;
+    default: break;
+  }
```

Comment:
get_json_object() has been merged  (https://github.com/rapidsai/cudf/blob/branch-0.19/cpp/src/strings/json/json_path.cu)
If we want to share parsing code (which I think is a good idea), I think this could be expressed as something like:

```
parse_result last_result = parse_result::SUCCESS;
while(1){
   PARSE_TRY(j_state.next_element());

  switch(j_state.cur_el_type){
  case ARRAY:
  case OBJECT:
      // step into the child elements (the fields of an object, or the elements of an array)
      json_state child(j_state);
      PARSE_TRY(child.child_element(NONE));
      while(last_result == parse_result::SUCCESS){
        // child.cur_el_name is your key, although for arrays I'm not sure what's supposed to happen here
        // as elements in an array don't have names.
        
        // to get the value, I think we would have to expose json_state::parse_value()
        // and make it take an output string_view parameter
        string_view value;
        if(child.parse_value(value) == parse_result::SUCCESS){
           ...
        }

        // next element
        PARSE_TRY(child.next_element());
      }
      break;

  default:
    // always an error?
    return parse_result::ERROR;
  }
}
```

Reply:
For sharing the code, we might do something like:

- move `parse_result`, `parser`, `json_output` (not sure about that one), `json_state` info `include/strings/detail/json.hpp` with implementations in `src/strings/json/json.cu`

Reply:
One thing to keep in mind - I expect the parser and json_state classes to become changed to work at the warp level.  That is to say, we will be marching the string data with an entire warp instead of one thread.  So those classes will end up also storing a thread id and whoever calls them (json_to_array) will have to also be warp-aware.